imap/sync_support.c

```diff
@@ -3284,6 +3185,202 @@ int sync_get_uniqueids(struct dlist *kin, struct sync_state *sstate)
 
 /* ====================================================================== */
 
+#ifdef USE_SIEVE
+static int list_cb(void *rock, struct sieve_data *sdata)
+{
+    strarray_t *list = (strarray_t *) rock;
+    size_t namelen = strlen(sdata->name);
+    char *name = xmalloc(namelen + 2); /* use trailing byte as active flag */
+
+    strcpy(name, sdata->name);
+    name[namelen+1] = sdata->isactive;
+
+    strarray_appendm(list, name);
+
+    return 0;
+}
+
+static int remove_cb(const char *sievedir, const char *fname,
+                     struct stat *sbuf, const char *link_target,
+                     void *rock)
+{
+    strarray_t *list = (strarray_t *) rock;
+    char path[PATH_MAX];
+    size_t namelen;
+
+    if (strarray_size(list)) {
+        if (S_ISREG(sbuf->st_mode)) {
+            const char *ext = strrchr(fname, '.');
+
+            if (ext && !strcmp(ext, BYTECODE_SUFFIX)) {
+                namelen = strlen(fname) - BYTECODE_SUFFIX_LEN;
+                snprintf(path, sizeof(path), "%.*s",
+                         (int) strlen(fname) - BYTECODE_SUFFIX_LEN, fname);
+                if (strarray_find(list, path, 0) >= 0) {
+                    /* Bytecode for an existing script - keep it */
+                    return SIEVEDIR_OK;
+                }
+            }
+        }
+        else if (S_ISLNK(sbuf->st_mode) && !strcmp(fname, DEFAULTBC_NAME)) {
+            namelen = strlen(link_target) - BYTECODE_SUFFIX_LEN;
+            snprintf(path, sizeof(path), "%.*s", (int) namelen, link_target);
+            if (strarray_find(list, path, 0) >= 0 && path[namelen+1] == 1) {
+                /* Active link for existing script - keep it */
+                return SIEVEDIR_OK;
+            }
+        }
```

Comment:
Ahh, here's the cause of the problem with XFER creating a "defaultbc" link during replication,  but not having it anymore later on!

Specifically, this line:

```
            if (strarray_find(list, path, 0) >= 0 && path[namelen+1] == 1) {
 ```

The problem is that `path[namelen+1]` is always `0`, because `path` was just created by `snprintf(..., "%.*s", (int) namelen, ...)` -- we will only ever include namelen bytes, and snprintf always 0-terminates, so the second half of this condition can never succeed -- and therefore we never recognise "Active link for existing script - keep it", and therefore we always throw away the defaultbc if we end up in this callback.

Assuming it's a typo, a quick change to make it compare `== 0` and my XFER test now passes... but at that point the second comparison is completely redundant because it will always be zero anyway, because of the snprintf.  Though the explicit comparison to `1` makes me feel like maybe something more subtle than I'm comprehending was intended here?  So I'm not sure that just removing the second condition is the right fix, but at least we've found the shape of the problem.

Reply:
I'm afk at until the morning but almost all of the code in my last commit was taken from master almost verbatim. I'm wondering if this problem also existed there or if the problem is in some changes that I made.

Reply:
master doesn't have `remove_cb()` at all.  I think this code was inherited from the older pre-uuid sieve-in-mailbox implementation, git blame says it's from around May

Reply:
OK.  So this issue was twofold:
1. Attempting to use a nasty hack of appending a trailing byte to the script name as an active flag in list_cb()
2. Checking the trailing byte of the wrong string in remove_cb()

I have removed this hack and instead always place the active script first in the array.  If no script is active and empty element will be in the first (zero) position.

Let me know if this solves the issue

Reply:
That seems to have fixed my XFER tests, great.  Now I'll try running them in a mixed version setup and see what happens